src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,820 @@
+// Copyright (c) Microsoft Corporation.
```

Comment:
Could you add custom messages to all assert and debug asserts that would dump the arguments?

`assert(a != b, "a={:?} b={:?}", a, b)`

Reply:
The only assertions in this file use the debug_assert_eq! macro, the documentation for which states:

> On panic, this macro will print the values of the expressions with their debug representations.

Is that not sufficient?

Reply:
No. In order to see the values the user has to set the `RUSTBACKTRACE=1` environment variable in advance, which may not be the case always, and even so it may not know how to print arguments.

Reply:
I'm not following.  The documentation for the debug_assert_eq! macro says it does what you're asking for, as I read it.  I don't understand what more you want me to do here.

Reply:
When the assertion failed (regardless of if it is a debug assert or not) the arguments do not get printed, unless the `RUST_BACKTRACE` environment variable is set.

I'm asking you to add these additional messages, anyways. This helps debugging code that is running without the aforementioned variable is set, especially when you hit a bug that is non-deterministic.

Reply:
Can you point to an example in the existing codebase where debug_assert_eq! or assert_eq! does what you're requesting?  All I'm seeing is hundreds of cases like mine.

Before we add a new stipulation on our code that will require changing potentially hundreds or thousands of lines of code, we should consider the impact, and whether there might be a better solution.  Such as redefining the assert macros to print the information you want even when RUSTBACKTRACE isn't set.  Or a workaround, like just having all of us set RUSTBACKTRACE=1 in our shell startup scripts (which would suffice for the debug versions of assert, since they're only meant for us anyway).

Reply:
I think we should separate this decision out in another PR after we discuss it as a group.